src/lsm/compaction.zig

```diff
@@ -311,293 +323,257 @@ pub fn CompactionType(
-            assert(compaction.data.ready or !compaction.data.ready);
+            if (!compaction.merge_iterator.?.empty()) {
+                compaction.cpu_merge();
+            } else {
+                compaction.cpu_merge_finish();
+            }
 
-            compaction.write_block_if_ready(&compaction.data, write_block_callback("data"));
-            compaction.write_block_if_ready(&compaction.filter, write_block_callback("filter"));
-            compaction.write_block_if_ready(&compaction.index, write_block_callback("index"));
+            // TODO Implement pacing here by deciding if we should do another compact_tick()
+            // instead of invoking the callback, using compaction.range.table_count as the heuristic.     
 
-            assert(!compaction.data.ready);
-            assert(!compaction.filter.ready);
-            assert(!compaction.index.ready);
+            const callback = compaction.callback.?;
+            compaction.callback = null;
+            callback(compaction);
         }
 
-        fn tick_cpu_merge(compaction: *Compaction) void {
+        fn cpu_merge(compaction: *Compaction) void {
+            // Ensure this is the result of a compact_tick() call that finished processing IO.
+            assert(compaction.status == .processing);
             assert(compaction.callback != null);
-            assert(compaction.ticks >= pipeline_tick_merge);
+            assert(compaction.io_pending == 0);
             assert(!compaction.merge_done);
-            assert(!compaction.merge_iterator.empty());
 
-            assert(!compaction.data.ready);
-            assert(!compaction.filter.ready);
-            assert(!compaction.index.ready);
+            // Ensure there are values to merge and that is it safe to do so.
+            assert(!compaction.merge_iterator.?.empty());
+            assert(!compaction.data.writable);
+            assert(!compaction.filter.writable);
+            assert(!compaction.index.writable);
 
+            // Build up the data/filter/index blocks with values merged from the read iterators.
             var tombstones_dropped: u32 = 0;
             while (!compaction.table_builder.data_block_full()) {
-                const value = compaction.merge_iterator.pop() orelse {
-                    compaction.assert_read_iterators_empty();
-                    break;
-                };
+                const value = compaction.merge_iterator.?.pop() orelse break;
                 if (compaction.drop_tombstones and tombstone(&value)) {
                     tombstones_dropped += 1;
                 } else {
                     compaction.table_builder.data_block_append(&value);
                 }
             }
 
-            if (compaction.table_builder.data_block_empty()) {
-                assert(compaction.drop_tombstones);
-                assert(tombstones_dropped > 0);
-            } else {
+            // Finalize the data block and prepare it to be written out.
+            blk: {
```

Comment:
What about:

```zig
if (compaction.table_builder.data_block_full() or
    compaction.table_builder.filter_block_full() or
    compaction.table_builder.index_block_full() or
    compaction.merge_iterator.?.empty())
{
    if (compaction.table_builder.data_block_empty()) {
        ... break :blk;
    }
```

— this way if `drop_tombstones` is `true` and iterator A has many tombstones, we won't create a bunch of tiny data blocks from values that _aren't_ dropped. Instead we'll collect them between ticks.